Reliable thread tracking using the monitor-fd feature in CLH

vineethrp opened the pull request.

Commit:
ch_monitor: Add support for cloud-hypervisor --monitor-fd option.

cloud-hypervisor has a new option to pass in a filedescriptor
through which we can listen for VM state change events. Add
support for --monitor-fd option. This would be used in later
patches to track VM's threads.

Signed-off-by: Vineeth Pillai <viremana@linux.microsoft.com>

Commit:
ch_monitor: Move the thread tracking to monitor object.

Now that we have a monitor notification from cloud-hypervisor,
it makes sense to move the thread tracking to monitor. Later
patches will add support for using the monitor framework from
cloud-hypervisor.

Signed-off-by: Vineeth Pillai <viremana@linux.microsoft.com>

Commit:
ch_monitor: Use the monitor framework from cloud-hypervisor

Use the events from cloud-hypervisor monitor framework to
handle VM lifecycle events and state changes

Signed-off-by: Vineeth Pillai <viremana@linux.microsoft.com>

Commit:
ch_monitor: Kill the cloud-hypervisor procees on monitor close

commit c857ac810c modified the VM creation by converting the
cloud-hypervisor as a daemon there by reparenting to init. So
we can no longer use virProcessAbort as it assumes that the
process to be killed is a child.

So use virProcessKillPainfully instead.

Signed-off-by: Vineeth Pillai <viremana@linux.microsoft.com>

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+        }
+
+        sz += ret;
+        events = virCHMonitorValidateEventsJSON(mon->buffer, sz, &incomplete);
```
Ah I see there is a lack of virjson supporting streaming, shame since that'd be nice for this usage instead of having to manually make these checks that rely heavily on the output format being correct.

vineethrp replied:
Yeah, we have to do this manually unfortunately!

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+                events++;
+        }
+        i++;
+    }
```
I think this is probably the right thing to do for now but this is really dependent on not getting transmission errors from cloud hypervisor. It might be worth extending virjson in the future to handle streaming so we don't need to do this kind of thing.

vineethrp replied:
While yajl supports streaming data, it does support only for a single document. In our case each event is one document and we may get multiple documents at the same time. So we would have to manually take care of that I suppose

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
```
This could get in an endless loop for the below so maybe make sure started is true before decrementing blocks or that blocks isn't already 0.
```
}
{}
```

bryteise replied:
Ah no sorry, this wouldn't get here with that kind of buffer.

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
 
     VIR_DEBUG("Monitor event loop thread starting");
 
+    if (VIR_ALLOC_N(mon->buffer, CH_MONITOR_BUFFER_SZ) < 0) {
```
Should switch to the glib allocation function here.

vineethrp replied:
Will do

vineethrp replied:
Fixed

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+        int ret;
+
+        ret = read(mon->monitor_fd, mon->buffer + sz,
+                   CH_MONITOR_BUFFER_SZ);
```
This could read past the end of the `mon->buffer` when `sz > 0`.

vineethrp replied:
Thanks for catching this. Will fix!

vineethrp replied:
Fixed

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+        sz += ret;
+        events = virCHMonitorValidateEventsJSON(mon->buffer, sz, &incomplete);
+        VIR_DEBUG("Monitor event(%ld):\n%s", sz, mon->buffer);
+    } while (virDomainObjIsActive(vm) && (events == 0 || incomplete));
```
So if `virCHMonitorValidateEventsJSON` returns `-1` (which is what `events` get set to) because the json in invalid, `incomplete` isn't set before it is checked here. This would cause a partial read of the monitor fd and likely continue to cause issues handling data from the buffer.

Maybe we should attempt to flush the buffer in case of invalid json?

vineethrp replied:
Correct, will fix!

vineethrp replied:
Fixed

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+        VIR_DEBUG("Monitor event(%ld):\n%s", sz, mon->buffer);
+    } while (virDomainObjIsActive(vm) && (events == 0 || incomplete));
+
+    return events ? virCHMonitorProcessEvents(mon, events) : 0;
```
I think `events` could be `-1` here and that would cause this to return 0 from the `virCHMonitorProcessEvents` call which seems maybe wrong.

vineethrp replied:
True.. I missed checking the sanity of events. Will fix in next iteration

vineethrp replied:
Fixed

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+            ret = -1;
+        }
+
+        *end = '{';
```
I'm a little confused by the restoration of `end` as a `{` since there shouldn't be any brackets left over with `virCHMonitorValidateEventsJSON` returning incomplete in that case I think.

bryteise replied:
Though I guess even if it wasn't the next value in the string could be whitespace and this would introduce an extra `{` to the string which would cause problems for `end_of_json`.

vineethrp replied:
This code is like a hard-coded parser for what cloud-hypervisor returns :-). It returns a strictly adjacent sets of JSON blocks now and just coded to that. But you are right, when we move this code to a more generice layer like virJSON, we need to handle for all edgecases like spaces, extra characters '\n' vs '\r\n' etc.

vineethrp replied:
Fixed all general cases

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +563,248 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+    int i = 0;
+
+    for (i = 0; i < events; i++) {
+        char *end = end_of_json(buf);
```
This could return past the last character of the buffer (`str + i` where `i == CH_MONITOR_BUFFER_SZ`) which is then written to so need to account for that by making a copy I think.

vineethrp replied:
Nice catch. Will fix!

vineethrp replied:
Fixed!

bryteise commented:
I'd really like to avoid having to do manual json poking if we can help it. There are a number of edge cases we might need to consider especially if we want to be robust against cloud-hypervisor sending malformed json.

vineethrp commented:
> I'd really like to avoid having to do manual json poking if we can help it. There are a number of edge cases we might need to consider especially if we want to be robust against cloud-hypervisor sending malformed json.

I had a quick pass on yajl and looks like it supports streaming only for a single document. In our case we may have multiple documents if there are multiple events happening in a row. Looks like we have to manually handle it for now.
We can surely workaround this by wrapping the it in a {} through the lifetime of a VM. We shall investigate on this as a separate effort.

I went through qemu code and it also does partial handling of the JSON document before handing it over to virjson. I hope its fine for ch driver also to have something similar

Commit:
ch_monitor: Save domain state after shutdown

Signed-off-by: Vineeth Pillai <viremana@linux.microsoft.com>

devimc commented on src/ch/ch_monitor.c:
```diff
@@ -560,6 +564,334 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+    int events = 0;
+    int i = 0;
+
+    /*
+     * Check if the message is a wellformed JSON.
+     */
+    while (i < sz) {
+        if (buffer[i] == '{') {
+            blocks++;
+        } else if (buffer[i] == '}'){
+            if (blocks == 0) {
+                VIR_WARN("JSON from monitor is not well-formed!");
+                return -1;
+            }
+            blocks--;
+            if (blocks == 0)
+                events++;
+        }
+        i++;
+    }
+    *incomplete = blocks != 0 ? true : false;
+    VIR_DEBUG("Total events received: %d, incomplete: %d",
+            events, *incomplete);
+
+    return events;
```
instead of parsing this manually, can you use `virJSONValueFromString` and `virJSONValueObjectKeysNumber` or `virJSONValueObjectForeachKeyValue` to get the number of events?

vineethrp replied:
The reason I did it this way is because, virJSONValueFromString expects a single JSON document. But we may have multiple documents received from monitor-fd due to multiple events being notified in succession. So we need to manually parse and split it into individual documents and then pass it on to virJSONValueFromString. Hope this clarifies

devimc replied:
how about converting these multiple documents into an array (adding `[` at the beginning and `]` at the end of the string) and parse it using `virJSONValueFromString`?

vineethrp replied:
I thought about that and infact had a discussion in a review comment in this PR. The problem is,  the event data from cloud-hypervisor is of streaming nature and currently virjson do not support streaming data(virJSONValueFromString calls yajl_complete_parse() assuming that the string is a complete json). So firstly we would need to support a streaming API in virjson. Trying to wrap in an array would need other complications as well. We would need to think more about it and to come up with a good design to have this feature.

devimc commented on src/ch/ch_monitor.c:
```diff
@@ -560,6 +564,334 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+    int blocks = 0;
+    int i = 0;
+
+    while ((i < len) && (!started || blocks > 0)) {
+        if (str[i] == '{') {
+            if (!started)
+                started = true;
+            blocks++;
+        } else if (str[i] == '}') {
+            blocks--;
+        }
+        i++;
+    }
+
+    return (i == len && blocks) ? NULL : str + i;
```
same here, can you use the `virJSONValue*` API instead ?

vineethrp replied:
Same  as above. We need to split into individual JSON documents and then pass it on to virJSONValueFromString one by one.

devimc reviewed (commented):
thanks @vineethrp 

I've left two comments/questions

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,294 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+    int events = 0;
+    size_t sz = 0;
+
+    memset(buf + sz, 0, max_sz);
```
I'm a little confused why sz is added here since it is always 0.

vineethrp replied:
oops sorry, it was a left over code that I missed removing. WIll remove it

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,294 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
-        virObjectUnlock(vm);
+        VIR_DEBUG("Reading events from monitor..");
+        if (virCHMonitorReadProcessEvents(mon) < 0)
+            VIR_WARN("Failed to process events from monitor!");
```
Should we consider this fatal? It seems like not being able to talk with the monitor (or the monitor behaving poorly) could be considered inoperable/unrecoverable.

vineethrp replied:
Similar to your comment below, I think its better to start of fresh rather than considering it fatal. Ultimately, what I would like to see is a two way channel for the monitor-fd(like qemu). There should be a mechanism to let cloud-hypervisor know that consumers cannot reliably get events and cloud-hypervisor should start fresh sending data. Not sure how feasible this would be, but will discuss with Rob about this. Probably like an HTTP API call also would do.

But I feel that instead of bailing out, we should reset the event monitor buffer and then start fresh(discard all data until the next starting '{' is detected). What do you think?

bryteise replied:
Yea, I like that approach. We just need to be sure we are keeping the buffer in a good state between calls in the eventloop while we are at it.

Needing to be sure we handle the cases of a full buffer being read but there not being a full event in the buffer so that we adjust everything so we can have the `{` at the start of the buffer for the next iteration of the eventloop.

Then we have to keep track of the case where we have a `{` at the start of the buffer and read the max size of our buffer but didn't see a full event. That case I'm not sure what the best way to handle is though, I'm open to suggestions.

vineethrp replied:
I have modified the code in such a way that no fatal situation exists after virCHMonitorReadProcessEvents returns

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,294 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+        }
+
+        sz += ret;
+        events = virCHMonitorValidateEventsJSON(mon->buffer,
```
Is trying to continue processing events valid if this gets a `-1` return? It only happens when the buffer has a `}` before a `{`. We could try and just ignore data in the buffer until we see the first `{` rather than return an error here though.

bryteise replied:
So basically just have `virCHMonitorValidateEventsJSON` not return `-1` and just get 0 events if it doesn't find a `{`.

bryteise replied:
Should cleanup the buffer too so we don't keep content before the `{`.

vineethrp replied:
I was thinking this was kinda fatal error as we are seeing a non wellformed JSON. But I get your point. Discarding bad data (emit a warning as well) and start fresh is a good idea. Will do that

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,294 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+     *    - The buffer is full and may have incomplete entries.
+     *
+     * If the buffer is full, virCHMonitorProcessEvents processes
+     * the completer events in the buffer and moves incomplete
```
`completer` -> `completed`

vineethrp replied:
Will fix

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,294 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+     *
+     * If the buffer is full, virCHMonitorProcessEvents processes
+     * the completer events in the buffer and moves incomplete
+     * entries to the start of the buffer and next read from teh pipe
```
`teh` -> `the`

vineethrp replied:
will fix

Commit:
ch_monitor: Fix array resize logic in iothread handling

Signed-off-by: Vineeth Pillai <viremana@linux.microsoft.com>

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,341 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
 
     VIR_DEBUG("Monitor event loop thread starting");
 
+    if (!(mon->buffer = g_malloc_n(sizeof(char), CH_MONITOR_BUFFER_SZ))) {
```
g_malloc_n should abort on failure so this check for return value shouldn't be needed.

vineethrp replied:
Ahh I didn't know that. Thanks for clarifying, will fix it

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,341 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+             * or in the system). Better to bail out.
+             */
+            VIR_ERROR("Failed to read monitor events!: %s", strerror(errno));
+            abort();
```
Thinking about this more, I'm not sure if this should abort since this could happen if a cloud-hypervisor process is killed for instance and the abort would cause the entire libvirt process to go down with a single vm. Perhaps just stopping the event loop, changing the vm's status to shutdown and returning would be correct behavior for this case?

vineethrp replied:
I think failure of read() system call on this pipe should be considered fatal. cloud-hypervisor process dying will not return error on read I think. It would be like encountering eof. read can return EAGAIN, EISDIR, EFAULT, EBADF and all these errors would be caused only due to a programming error in our driver. It can return EIO which could be caused wither by a programming error or a deep kernel level problem. Other than that EINTR can happen and we take care of that.

I am approaching it as a failfast mechanism where we detect the bug in our code fast by aborting libvirtd and then fix the problem in code fast. I agree EIO is out of our hands and probably should handled like what you mentioned. But for rest of the cases I think its better to crash libvirtd so that we as developers get to know the problem if it happens.

Also killing of the libvirtd should not kill the VMs. Since we run each VM as a daemon and parented to init, VMs should run unaffected. Restarting libvirtd would bring everything back to normal I guess.

What do you feel?

bryteise replied:
Ah yes you are right, it would hit an EOF since it can read from a closed pipe up to however much data was left over.

For the EOF case, I'm not sure if we shouldn't just stop the loop as we just are adding wake-ups to the system every second otherwise but I don't think it makes much difference so I'm fine with the code as is.

vineethrp replied:
EOF case is already taken care off I guess(if virDomainObjIsActive(vm) returns false when cloud-hypervisor gets killed). As of now, Normall kill command (SIGTERM) works as expected. But SIGKILL doesn't notify libvirt. We would need to fix this and then the loop will be exited with out any extra effort. 
We have an action item here to make libvirt aware of cloud-hypervisor abnormal exits. Filed and issue here: https://github.com/cloud-hypervisor/libvirt/issues/47

bryteise replied:
RIght, of course. Thank you so much for going through all this. Quite a lot of details to get right X(.

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+                int invalid_chars = json_start - scan_start;
+                VIR_WARN("invalid json or white chars in buffer: %.*s",
+                         invalid_chars, scan_start);
+                memmove(scan_start, json_start, invalid_chars);
```
I'm not so sure about this `memmove`. It copies `invalid_chars` number of items from the valid part of the string over the invalid part which doesn't seem useful. What about the following:
```
memmove(buffer, json_start, sz - invalid_chars);
memset(buffer + sz - invalid_chars, 0, invalid_chars);
json_start = 0;
scan_start = 0;
i = 0;
sz -= invalid_chars;
```
This copies the contents from the start of the json string to the end of the buffer (skipping over the invalid characters essentially) to the start of the buffer, then zeroes out the end of the buffer for however many invalid characters there were (and then resets the various tracking offsets).

What do you think about this approach?

vineethrp replied:
Sorry, what I wanted was actually move rest of the buffer from json_start to scan_start. The exact idea that you proposed above was my intention, but I messed it up ;-)

Moving to start of the buffer might not always be correct. The invalid chars can be between two valid json blocks as well. So we need to move only those areas. A slightly modified version of your proposal:
```
memmove(scan_start, json_start, sz - i);
memset(scan_start + sz - i, 0, invalid_chars);
json_start = scan_start;
i -= invalid_chars;
sz -= invalid_chars;
```
How about this?

bryteise replied:
Ahh you are certainly correct, I wasn't thinking about the invalid characters between full blocks. Your proposal looks right to me.

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+        i++;
+    }
+
+    return (i == len && blocks) ? NULL : str + i;
```
Couldn't we have a full buffer that ends perfectly at a closing `}`. In that case this would return `str + i` as blocks would be 0 but `str + i` would be the location one past the end of the buffer and wouldn't that be an overflow (as we read/write to the that location in `virCHMonitorProcessEvents`)?

vineethrp replied:
yes you are right. Will fix!

vineethrp replied:
Hmm on going through the code again, I think we have a bigger problem. virJSONValueFromString requires that EOL('\0') to identify the end of the document. So we would need to grow the buffer to get it processed correctly. Thats not a scalable solution. So we would need to make sure that the JSON doc size should be atmost one less than the buf size. We shall discuss this during tomorrow's meeting and have the protocol guarentee that, the JSON doc is with in size limits. And we wiil consider a JSON doc invalid even if its more than he protocol size

vineethrp replied:
But there is another probability that we hit a smaller JSON doc which happens to be right at the end of the buffer. I will take care of this in the code.

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+     * If the buffer still has incomplete data, lets
+     * push it to the beginning.
+     */
+    sz = buf - mon->buffer;
```
I think `sz` is the amount of data that was processed (`mon->buffer` is the start and `buf` is the location after the last json event that was processed).

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+     */
+    sz = buf - mon->buffer;
+    if (sz < mon->buf_fill_sz) {
+        mon->buf_offset = mon->buf_fill_sz - sz;
```
And then here `mon->buf_offset` is set to the location past the remainder of the data in the buffer.

bryteise commented on src/ch/ch_monitor.c:
```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+    sz = buf - mon->buffer;
+    if (sz < mon->buf_fill_sz) {
+        mon->buf_offset = mon->buf_fill_sz - sz;
+        memmove(mon->buffer, buf, sz);
```
I think `sz` is the amount of data that was processed (`mon->buffer` is the start and `buf` is the location after the last json event that was processed).

And then here `mon->buf_offset` is set to what will be the index past the number of bytes we haven't read in the buffer.

But then we move `sz` which isn't how much was left (which I think is `mon->buf_fill_sz - sz`).

vineethrp replied:
Ahh yes you are right. Thanks for finding this. Will fix.

Commit:
ch_monitor: Address review comments

Addressing the review comments:
https://github.com/cloud-hypervisor/libvirt/pull/45#discussion_r584421793
https://github.com/cloud-hypervisor/libvirt/pull/45#discussion_r584418416
https://github.com/cloud-hypervisor/libvirt/pull/45#discussion_r584407145

Signed-off-by: Vineeth Pillai <viremana@linux.microsoft.com>
